[BACKEND] Pipeliner refactoring

ThomasRaoux opened the pull request:
Refactor the pipeliner pass in order to make it more generic. The main change is that the pipeliner is now broken into 2 pieces one calculating a modulo schedule and create async ops based on the IR and an expander that will generate the pipelined IR based on the modulo schedule.
The advantage of separating the two pieces is that it will allow us to create different schedule without having to change the expander and it will allow for more complex schedules.
For now the schedule generated for matmul case matches rougly the schedule picked by the previous pipeliner in order to avoid changes.

This also creates a different sequence of insert/extract slice for the alloc. We should probably change shared alloc to use memory semantic.

Jokeren commented:
What does the last sentence mean?

> We should probably change shared alloc to use memory semantic.

ThomasRaoux commented:
> What does the last sentence mean?
> 
> > We should probably change shared alloc to use memory semantic.

I started a thread with few people about this but the main idea is to change the sequence:
```
%10 = triton_gpu.alloc_tensor : tensor<2x128x32xf16, #shared>
 %14 = triton_gpu.insert_slice_async %4, %10, %c0_i32, %13 {axis = 0 : i32, cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<128x32x!tt.ptr<f16, 1>, #blocked1> -> tensor<2x128x32xf16, #shared>
 %25 = triton_gpu.extract_slice %14[%c0_i32, 0, 0] [1, 128, 32] [1, 1, 1] : tensor<2x128x32xf16, #shared> to tensor<128x32xf16, #shared>
 ```
so that `insert_slice_async` doesn't return a new tensor to something like this:
```
 %10 = triton_gpu.alloc_tensor : !tt.ptr<tensor<2x128x32xf16, #shared>, 3>
 triton_gpu.copy_slice %4, %10, %c0_i32, %13 {axis = 0 : i32, cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<128x32x!tt.ptr<f16, 1>, #blocked1>, !tt.ptr<tensor<2x128x32xf16, #shared>, 3> -> tensor<2x128x32xf16, #shared>
 %25 = triton_gpu.get_slice %10[%c0_i32, 0, 0] [1, 128, 32] [1, 1, 1] : !tt.ptr<tensor<2x128x32xf16, #shared>, 3> to tensor<128x32xf16, #shared>
```

zahimoud commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.h:
```diff
@@ -0,0 +1,101 @@
+  // the predicated equivalent and return it, or return nullptr if the
+  // predication is impossible. In the latter case, pipelining will fail and
+  // may leave IR in a partially transformed state.
+  using PredicateOpFn =
```
Should this be `PredicateOpFnType` ?

ThomasRaoux replied:
Changed

zahimoud commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+  // 2. Convert the loads into async loads and create the allocs.
+  createAsynOps(forOp, loads, numStages, hasMMAV3);
+
+  // 3. rewrite the loop using the given schedule, this part of the
```
Is this comment true ?

ThomasRaoux replied:
yes this comment was outdated. Fixed it.

zahimoud commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+      if (!candidate)
+        continue;
+      Value dotOperand = loadDotOperand(loadOp, hasMMAV3);
+      if (!dotOperand)
```
Do we only pipeline dotOp loads ?

zahimoud replied:
I guess this is just to be on par with the current pipeliner, but do you think there will be an issue relying on LoadDotOp struct if we want to extend for other loads ?

ThomasRaoux replied:
Right this is to match existing level of support. We can pipeline whatever we want as long as we pick a schedule for it and generate the async wait correct. 

>do you think there will be an issue relying on LoadDotOp struct if we want to extend for other loads ?
The goal of the refactor is to allow this kind of cases so I don't expect any issue but there may be some rough edges once we start doing it and we should improve things based on that.

zahimoud replied:
Nevermind, I think it should be fine

zahimoud commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+
+static void setWaitNum(Operation *op,
+                       mlir::triton::PipeliningOption::PipelinerPart part,
+                       unsigned iteration, unsigned numLoadsInStage) {
```
What is `iteration` supposed to be used for ?

ThomasRaoux replied:
it tells you the iteration of the prologue/epilogue, this allows you to know what "version" of the operation you are getting.

zahimoud replied:
ohh i see, so we would special case based on the iteration ?

ThomasRaoux replied:
right for instance if you have a wait in the prologue you might have to wait for different numbers of copy

zahimoud replied:
yes, now I get it. I was confused that `iteration` wasn't used

zahimoud commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp:
```diff
@@ -0,0 +1,683 @@
+        assert(newOp && "failed to predicate op.");
+      }
+      if (annotateFn)
+        annotateFn(newOp, triton::PipeliningOption::PipelinerPart::Prologue, i);
```
How would this actually work if we have chained loads ? the `wait` is always going to be this: `unsigned numLoadsInStage = (numStages - 2) * loads.size();`, but in the prologue we might need some async loads to be done so that we can launch the dependent async loads, so basically we need `wait 0`

zahimoud replied:
I guess this is related to my previous comment, basically using `iteration` would let us set the wait num differently ?

ThomasRaoux replied:
Right that's why we can know what version we get in the annotate lambda so that we can set the right value on the wait

zahimoud replied:
Makes sense!

zahimoud commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+// create the schedule for a matmul loop. This is ad hoc based on how we know
+// matmul loops should be pipelined and is not a generic scheduler.
+static std::vector<std::pair<Operation *, unsigned>>
+createSchedule(scf::ForOp forOp, int numStages) {
```
Do you think we can still let the user specify some sort of high level schedule in the kernel implementation ? The schedule in here is created at a lower level (after we create async ops), so a high level schedule wouldn't map 1:1 to the ops at this level.

ThomasRaoux replied:
Right, I think we can still do it but it would be some very coarse level information. For instance if we get a chain of long latency op we could have hints telling in what stage to put each middle op. For just a load+dot loop thee is probably not much to expose to user.

Jokeren commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp:
```diff
@@ -0,0 +1,683 @@
+    Location loc = forOp.getLoc();
+    Type t = lb.getType();
+    // newLastIter = lb + step * ((((ub - 1) - lb) / step) - i)
+    Value newlastIter = rewriter.create<arith::AddIOp>(
```
A bit worries about the readability here

ThomasRaoux replied:
indeed, this is pretty terrible. Refactored it, hopefully it is a bit better.

Jokeren commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp:
```diff
@@ -0,0 +1,683 @@
+//
+//===----------------------------------------------------------------------===//
+
+// Fork of upstream pipeliner. This will be merged upstream once things are
```
I'm not quite clear about the exact code changes comparing this file with the upstream pipeline pass. Can you pin me when you prepare to commit changes to LLVM/MLIR?

ThomasRaoux replied:
Definitely, I'll include you on the PRs.

Jokeren commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+#include "PipelineExpander.h"
```
I suppose AMD could have an independent file for pipelining and set up proper pipelining options?

ThomasRaoux replied:
Yes that's the idea. I think we should also be able to extra common helpers to make it easier to create schedules but I don't have a clear idea of which helpers would be useful at this point.

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+  auto commmit = builder.create<ttg::AsyncCommitGroupOp>(loc);
+
+  // Extract part.
+  if (insertWait)
```
Mostly because I want the wait to be before any of the extract and since I don't want to accidently schedule to late I insert it here but I realize this is not great. I moved this outside.

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+    return;
+  for (Value operand : op->getOperands()) {
+    Value v = operand;
+    int distance = 0;
```
removed.

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+  }
+
+  // Find depenencies with distance of 1.
+  SmallVector<Operation *> distance1Arith;
```
good point, fixed the name.

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -0,0 +1,691 @@
+      }
+    }
+  }
+  // Keep loads at a distance of 1 schedule the rest in stage 0.
```
That's a better comment indeed.

ThomasRaoux commented:
Thanks for reviewing. I added support for TMA and addressed the comments. Please take another look, if everybody is happy with it, it should be good to go

Jokeren reviewed (approved):
LGTM, thanks!
